cpp/src/arrow/adapters/orc/adapter.cc

```diff
@@ -545,24 +542,16 @@ class ORCFileReader::Impl {
     if (length == 0) {
       return Status::OK();
     }
-    int64_t start = builder->length();
 
     const uint8_t* valid_bytes = nullptr;
     if (batch->hasNulls) {
       valid_bytes = reinterpret_cast<const uint8_t*>(batch->notNull.data()) + offset;
     }
-    RETURN_NOT_OK(builder->AppendNulls(valid_bytes, length));
-
     const int64_t* source = batch->data.data() + offset;
-    uint8_t* target = reinterpret_cast<uint8_t*>(builder->data()->mutable_data());
+    std::vector<uint8_t> converted(source, source + length);
+
+    RETURN_NOT_OK(builder->AppendValues(converted.data(), length, valid_bytes));
```

Comment:
Hmm... I think it's a bit problematic that we seem to be introducing a second data copy here (first from `source` to `converted`, then from `converted` to the builder). Can we avoid the second copy without using `builder->data()`? Otherwise we may have to undeprecate the latter.

Reply:
Note we don't have any unit tests for this code path so have to be extra careful